Sources/GRPCHTTP2Core/Server/GRPCServerStreamHandler.swift

```diff
@@ -0,0 +1,218 @@
+  }
+
+  func handlerRemoved(context: ChannelHandlerContext) {
+    self.stateMachine.tearDown()
+  }
+}
+
+// - MARK: ChannelOutboundHandler
+
+@available(macOS 13.0, iOS 16.0, watchOS 9.0, tvOS 16.0, *)
+extension GRPCServerStreamHandler {
+  func write(context: ChannelHandlerContext, data: NIOAny, promise: EventLoopPromise<Void>?) {
+    let frame = self.unwrapOutboundIn(data)
+    switch frame {
+    case .metadata(let metadata):
+      do {
+        let headers = try self.stateMachine.send(metadata: metadata)
+        context.write(self.wrapOutboundOut(.headers(.init(headers: headers))), promise: nil)
+        self.flushPending = true
+        // TODO: move the promise handling into the state machine
+        promise?.succeed()
+      } catch {
+        context.fireErrorCaught(error)
+        // TODO: move the promise handling into the state machine
+        promise?.fail(error)
+      }
+
+    case .message(let message):
+      do {
+        try self.stateMachine.send(message: message, endStream: false)
+        // TODO: move the promise handling into the state machine
+        promise?.succeed()
+      } catch {
+        context.fireErrorCaught(error)
+        // TODO: move the promise handling into the state machine
+        promise?.fail(error)
+      }
+
+    case .status(let status, let metadata):
+      do {
+        let headers = try self.stateMachine.send(status: status, metadata: metadata)
+        let response = HTTP2Frame.FramePayload.headers(.init(headers: headers, endStream: true))
+        self.pendingTrailers = response
+        // TODO: move the promise handling into the state machine
+        promise?.succeed()
+      } catch {
+        context.fireErrorCaught(error)
+        // TODO: move the promise handling into the state machine
+        promise?.fail(error)
+      }
+    }
+  }
+
+  func flush(context: ChannelHandlerContext) {
+    do {
+      if self.isReading {
+        // We don't want to flush yet if we're still in a read loop.
+        return
+      }
+
+      var nextOutboundMessage = try self.stateMachine.nextOutboundMessage()
+      while case .sendMessage(let byteBuffer) = nextOutboundMessage {
+        context.write(
+          self.wrapOutboundOut(.data(.init(data: .byteBuffer(byteBuffer)))),
+          promise: nil
+        )
+        self.flushPending = true
+        nextOutboundMessage = try self.stateMachine.nextOutboundMessage()
+      }
+
+      switch nextOutboundMessage {
+      case .noMoreMessages:
+        if let pendingTrailers = self.pendingTrailers {
+          context.write(self.wrapOutboundOut(pendingTrailers), promise: nil)
+          self.flushPending = true
+          self.pendingTrailers = nil
+        }
+      case .awaitMoreMessages:
+        ()
+      case .sendMessage:
+        preconditionFailure("This isn't possible: we'd still be inside the while loop.")
+      }
+
+      if self.flushPending {
+        context.flush()
+        self.flushPending = false
```

Comment:
General rule of thumb: always update state _before_ making an outcall. The outcall may re-entrantly call this handler which could update state in some other way, which would then be trampled over when the stack unwinds.

```suggestion
        self.flushPending = false
        context.flush()
```